README


This archive contains the distribution Locale-Maketext-Lexicon-DB,
version 1.141190:

  Dynamically load lexicon from a database table

This software is copyright (c) 2014 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




CHANGES
==================================================
Changes from 2013-04-29 00:00:00 +0000 to present.
==================================================

---------------------------------------------
version 1.141190 at 2014-04-29 18:58:54 +0000
---------------------------------------------

  Change: 485e56d9b6fb8f1378b86a357abe499b5bf08a33
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-04-29 19:55:39 +0000

    Fix tests, cache key and change level of logging

    - Skip memached tests if cannot connect - Fix the format of the cache
    key - Changed logging level to trace across the board 

  Change: 92982ab5195f339a430e18949f68694507ccfa11
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-01-20 13:17:33 +0000

    Fix test to not fail if memcached not installed 

  Change: a992581811c72bfe3f682f67974b820bd11be28f
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-01-10 17:41:50 +0000

    Fix tests and pod 

  Change: 1744f76db4c91ec82617b75c3c906043096f1174
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-01-09 16:29:05 +0000

    Update pod 

  Change: 24c62cda125fb4b47ed37f21ee9080f4d59aae5a
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-01-09 11:20:38 +0000

    Update test and pod 

  Change: 93648e74dbf53ae9df1520018d4d5865f4df4383
  Author: Pete Smith <pete.smith@net-a-porter.com>
  Date : 2014-01-08 16:08:07 +0000

    Initial commit 

================
End of releases.
================


lib/Locale/Maketext/Lexicon/DB.pm
package Locale::Maketext::Lexicon::DB;
{
  $Locale::Maketext::Lexicon::DB::VERSION = '1.141190';
}
# ABSTRACT: Dynamically load lexicon from a database table

use Locale::Maketext::Lexicon::DB::Handle;
use Moose;
use namespace::autoclean;
use Locale::Maketext 1.22;
use Log::Log4perl qw(:easy);


has dbh => (
    is          => 'ro',
    isa         => 'Object',
    required    => 1,
);

has cache => (
    is          => 'ro',
    isa         => 'Object',
    predicate   => 'has_cache',
);

has cache_expiry_seconds => (
    is          => 'ro',
    isa         => 'Int',
    default     => 60 * 5,
);

has lex => (
    is          => 'ro',
    isa         => 'Str',
    required    => 1,
);

has auto => (
    is          => 'ro',
    isa         => 'Bool',
    default     => 0,
);

has language_mappings => (
    is          => 'ro',
    isa         => 'HashRef[ArrayRef]',
    required    => 1,
);


{
    my $instance;

    sub get_handle {
        my $class = shift;
        my @requested_langs = @_;

        $instance ||= $class->new;

        @requested_langs = Locale::Maketext->_ambient_langprefs
            unless @requested_langs;

        TRACE('Languages asked for: ' . join (', ', @requested_langs));

        my $langs = [];
        for (@requested_langs) {
            if (defined $class->new->language_mappings->{ lc $_ }) {
                $langs = $class->new->language_mappings->{ lc $_ };
                last;
            }
        }

        TRACE('Lexicon will be searched for languages: ' . join(', ', @{ $langs }) );

        return Locale::Maketext::Lexicon::DB::Handle->new(
            _parent => $instance,
            langs   => $langs
        );
    }
}


sub clear_cache {
    my $class = shift;

    my $self = $class->new;

    if (defined $self->cache) {
        for (values %{ $self->language_mappings }) {
            $self->cache->delete( $self->_cache_key_for_langs($_) );
        }

        return 1;
    }

    return;
}

sub _cache_key_for_langs {
    my $self = shift;

    return join(
        '.',
        'lexicon',
        $self->lex,
        @{ shift() }
    )
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Locale::Maketext::Lexicon::DB - Dynamically load lexicon from a database table

=head1 VERSION

version 1.141190

=head1 SYNOPSIS

    package MyApp::Maketext;

    use Moose;
    use DBI;
    use Cache::Memcached::Fast;

    BEGIN { extends 'Locale::Maketext::Lexicon::DB'; }

    has '+dbh' => (
        builder => '_build_dbh',
    );

    sub _build_dbh {
        my $self = shift;

        return DBI->connect( ... );
    }

    has '+cache' => (
        builder => '_build_cache',
    );

    sub _build_cache {
        my $self = shift;

        return Cache::Memcached::Fast->new({
            servers => [ ... ],
        });
    }

    has '+cache_expiry_seconds' => (
        default => 3_600,
    );

    has '+lex' => (
        default => 'myapp',
    );

    has '+auto' => (
        default => 1,
    );

    has '+language_mappings' => (
        default => sub {
            {
                en_gb   => [qw(en_gb en)],
                en_us   => [qw(en_us en)],
                en      => [qw(en)],
            }
        },
    );


    package main;

    my $handle = MyApp::Maketext->get_handle('en_gb');
    # or, to use the environment to get the language from locale settings:
    # my $handle = MyApp::Maketext->get_handle;

    print $handle->maketext('ui.homepage.title', $name);

=head1 DESCRIPTION

This module enables you to crate a L<Locale::Maketext> lexicon in your database. The lexicon is
compiled when C<get_handle> is called on the class. If a cache is defined then the lexicon is
retrieved from the cache instead of hitting the database each time. A class method is provided to
clear the cache i.e. to invalidate it if the lexicon in the DB changes.

=head1 METHODS

=head2 get_handle ([@languages])

Returns a L<Locale::Maketext::Lexicon::DB::Handle> for this lexicon. if C<@languages> are not
supplied then inspects the environment to get the set locale.

=head2 clear_cache

Clears the cache (if set) for this lexicon. Used to invalidate the cache if the database has
changed.

=head1 DATABASE TABLE

Your database should be in this format (this DDL is for SQLite).

    CREATE TABLE lexicon (
        id INTEGER PRIMARY KEY NOT NULL,
        lang VARCHAR NOT NULL,
        lex VARCHAR NOT NULL,
        lex_key TEXT NOT NULL,
        lex_value TEXT NOT NULL
    );

=over 4

=item id

The primary key for the table

=item lang

The locale string for the language for this entry

=item lex

A key to identify the entire lexicon in the table. This enables you to set define more than one
lexicon in the table

=item lex_key

The key for the lexicon entry. This is the value passed to the C<maketext> method on the handle

=item lex_value

The value for the lexicon entry

=back

=head1 AUTHOR

Pete Smith <pete@cubabit.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Locale/Maketext/Lexicon/DB/Handle.pm
package Locale::Maketext::Lexicon::DB::Handle;
{
  $Locale::Maketext::Lexicon::DB::Handle::VERSION = '1.141190';
}
# ABSTRACT: Maketext handle for L<Locale::Maketext::Lexicon::DB>

use Moose;
use namespace::autoclean;
use Locale::Maketext;
use Carp;
use Data::Dumper;
use Log::Log4perl qw(:easy);

has langs => (
    is          => 'ro',
    isa         => 'ArrayRef',
    required    => 1,
);

has _parent => (
    is          => 'ro',
    isa         => 'Object',
    required    => 1,
);

sub _lexicon {
    my $self = shift;

    my $lexicon = {};
    my $cache_key;

    if ($self->_parent->has_cache) {
        $cache_key = $self->_parent->_cache_key_for_langs( $self->langs );

        local $Storable::Eval = 1;
        $lexicon = $self->_parent->cache->get($cache_key);

        TRACE('Retrieved lexicon from cache')
            if (defined $self->_parent->cache);
    }

    unless (keys %{ $lexicon }) {
        TRACE('Hitting database for lexicon');

        my $dbh = $self->_parent->dbh;

        for my $lang (@{ $self->langs }) {
            TRACE('Getting lexicon entries for language ' . $lang);

            my $lexicon_st = $dbh->prepare(q{
                SELECT *
                FROM lexicon
                WHERE lex = ?
                AND lang = ?
            });
            $lexicon_st->execute($self->_parent->lex, $lang);

            while (my $lex_entry = $lexicon_st->fetchrow_hashref) {
                my $key     = $lex_entry->{lex_key};
                next if defined $lexicon->{ $key };

                my $value   = $lex_entry->{lex_value};
                $value      =~ tr/\r//d;
                $value      =~ s/\n/\\n/g;

                $lexicon->{ $key } = Locale::Maketext->_compile($value);
            }
        }

        if ($self->_parent->has_cache and keys %{ $lexicon }) {
            local $Storable::Deparse = 1;
            TRACE('Storing lexicon in cache');
            $self->_parent->cache->set($cache_key => $lexicon, $self->_parent->cache_expiry_seconds);
        }
    }

    TRACE('Compiled lexicon is: ' . Dumper $lexicon);

    return $lexicon;
}


sub maketext {
    my $self = shift;
    my $key = shift;

    my $value = $self->_lexicon->{ $key };

    unless (defined $value) {
        if ($self->_parent->auto) {
            $value = Locale::Maketext->_compile($key);
        }
        else {
            croak $key . ' not found in lexicon';
        }
    }

    if (ref $value eq 'SCALAR') {
        return ${ $value };
    }

    TRACE('Returning maketext entry for key ' . $key);

    return $value->( 'Locale::Maketext', @_ );
}


__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Locale::Maketext::Lexicon::DB::Handle - Maketext handle for L<Locale::Maketext::Lexicon::DB>

=head1 VERSION

version 1.141190

=head1 METHODS

=head2 maketext ($key)

Return the compiled string represented by C<$key>. If C<$key> does not exist, an exception is
raised, unless C<auto> is set in which case C<$key> is returned.

=head1 SEE ALSO

L<Locale::Maketext::Lexicon::DB>

=head1 AUTHOR

Pete Smith <pete@cubabit.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Dynamically load lexicon from a database table",
  "AUTHOR" => "Pete Smith <pete\@cubabit.net>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Locale-Maketext-Lexicon-DB",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Locale::Maketext::Lexicon::DB",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Locale::Maketext" => "1.22",
    "Log::Log4perl" => 0,
    "Moose" => 0,
    "namespace::autoclean" => 0
  },
  "TEST_REQUIRES" => {
    "Cache::Memcached::Fast" => 0,
    "DBD::SQLite" => 0,
    "DBI" => 0,
    "File::Temp" => 0,
    "Test::Memcached" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.141190",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Cache::Memcached::Fast" => 0,
  "Carp" => 0,
  "DBD::SQLite" => 0,
  "DBI" => 0,
  "Data::Dumper" => 0,
  "File::Temp" => 0,
  "Locale::Maketext" => "1.22",
  "Log::Log4perl" => 0,
  "Moose" => 0,
  "Test::Memcached" => 0,
  "Test::More" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



